### PR TITLE
feat: introduce estimate size api for executors

### LINKED ISSUE #8957: Implement `EstimateSize` for data structures
We need to implement trait `EstimateSize` for data structures like wrapped `BtreeMap` and `HashSet`.

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- add `EstimateSize` trait API to calculate the operator memory usage.

TODO: Need to resolve #8957 in our codes.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
~- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/collection/estimate_size.rs [hunk 1]
```diff
@@ -37,27 +40,63 @@ impl EstimateSize for FixedBitSet {
...
+            0
+        }
+    }
+}
+
+/// SAFETY: `Bytes` can store a pointer in some cases, that may cause the size
+/// of a `Bytes` be calculated more than one and when memory stats is larger than the real value.
+impl EstimateSize for Bytes {
+    fn estimated_heap_size(&self) -> usize {
+        self.len()
```
Comment (liurenjie1024) [inline src/common/src/collection/estimate_size.rs:74]: Does it has sth similar to `capacity`?

File: src/common/src/collection/estimate_size.rs [hunk 2]
```diff
@@ -37,27 +40,63 @@ impl EstimateSize for FixedBitSet {
...
+    }
+}
+
+// FIXME: implement a wrapper structure for `HashSet` that impl `EstimateSize`
+// https://github.com/risingwavelabs/risingwave/issues/8957
+impl<T: EstimateSize> EstimateSize for HashSet<T> {
+    fn estimated_heap_size(&self) -> usize {
+        0
```
Comment (liurenjie1024) [inline src/common/src/collection/estimate_size.rs:82]: Why this is zero?

**Review by liurenjie1024 — COMMENTED** [review]
Generally fine with this. But mutable data structures like `TopNCache`, the main part is the maintain it's estimated heap size during eviction and insertion.

File: src/common/src/collection/estimate_size.rs (hunk 2 above)
Comment (yuhao-su) [reply src/common/src/collection/estimate_size.rs:82]: I let it to be fixed. A issue was noted above.

File: src/common/src/collection/estimate_size.rs (hunk 1 above)
Comment (yuhao-su) [reply src/common/src/collection/estimate_size.rs:74]: No. You can think it as len is always that same as capacity

**Comment by yuhao-su** [comment]
> But mutable data structures like TopNCache

Let me add a similar interface for `ExecutorCache`

File: src/stream/src/executor/hash_agg.rs [hunk 3]
```diff
@@ -547,20 +546,24 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
...
+                // from `group_change_set` which is a set.
+                //
+                // 2. `MutGuard` should not be sent to other tasks.
+                let mut agg_group = unsafe { agg_group.as_mut_guard() };
+                async move {
+                    // Get agg outputs and build change.
+                    let curr_outputs = agg_group.get_outputs(storages).await?;
+                    let change = agg_group.build_change(curr_outputs);
+                    Ok::<_, StreamExecutorError>(change)
+                }
```
Comment (st1page) [inline src/stream/src/executor/hash_agg.rs:566]: :thinking:

Comment (yuhao-su) [reply src/stream/src/executor/hash_agg.rs:566]: ?

File: src/stream/src/executor/aggregation/distinct.rs [hunk 4]
```diff
@@ -74,8 +74,11 @@ impl<S: StateStore> ColumnDeduplicater<S> {
...
             let compacted_key = CompactedRow::from(&key); // TODO(rc): is it necessary to avoid recomputing here?
-            let counts = if let Some(counts) = self.cache.get_mut(&compacted_key) {
-                counts
+
+            // TODO(yuhao): avoid this `contains`.
+            // https://github.com/risingwavelabs/risingwave/issues/9233
+            let mut counts = if self.cache.contains(&compacted_key) {
+                self.cache.get_mut(&compacted_key).unwrap()
             } else {
                 // load from table into the cache
```
Comment (st1page) [inline src/stream/src/executor/aggregation/distinct.rs:83]: can we write this here? :thinking: 
```suggestion
    let mut cache = UNSAFE { cache };
```

**Review by liurenjie1024 — APPROVED** [review]

**Review by BugenZhao — APPROVED** [review]

**Pull request merged**